lib/internal/fs/promises.js

```diff
@@ -316,25 +315,46 @@ async function readFileHandle(filehandle, options) {
   if (size > kIoMaxLength)
     throw new ERR_FS_FILE_TOO_LARGE(size);
 
-  const chunks = [];
-  let isFirstChunk = true;
-  const firstChunkSize = size === 0 ? kReadFileMaxChunkSize : size;
-  const chunkSize = MathMin(firstChunkSize, kReadFileMaxChunkSize);
   let endOfFile = false;
+  let totalRead = 0;
+  const noSize = size === 0;
+  const buffers = [];
+  const fullBuffer = noSize ? undefined : Buffer.allocUnsafeSlow(size);
   do {
     if (signal?.aborted) {
       throw lazyDOMException('The operation was aborted', 'AbortError');
     }
-    const buf = Buffer.alloc(isFirstChunk ? firstChunkSize : chunkSize);
-    const { bytesRead, buffer } =
-      await read(filehandle, buf, 0, buf.length, -1);
-    endOfFile = bytesRead === 0;
-    if (bytesRead > 0)
-      ArrayPrototypePush(chunks, buffer.slice(0, bytesRead));
-    isFirstChunk = false;
+    let buffer;
+    let offset;
+    let length;
+    if (noSize) {
+      buffer = Buffer.allocUnsafeSlow(kReadFileUnknownBufferLength);
+      offset = 0;
+      length = kReadFileUnknownBufferLength;
+    } else {
+      buffer = fullBuffer;
+      offset = totalRead;
+      length = MathMin(size - totalRead, kReadFileBufferLength);
+    }
```

Comment:
Note that this is a behavioral change for files that are being appended to while they are being read. I think that *should* be okay, because ultimately there are no guarantees for the relative timing of the two operations, but it might be something to keep in mind.

Reply:
Yes, you're correct and I should've mentioned it in the PR but forgot. This is aligning the behaviour with the cb version, so I thought that it would be OK.

Reply:
Yeah, I think that’s okay :+1: